### PR TITLE
[MLIR][TORCH] Add E2E support for aten.avg_pool2d op

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit adds lowering of `aten.avg_pool2d` op.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

### CODE REVIEW

**Comment by vivekkhandelwal1** [comment]
Hi, I am getting the following error for the e2e test case:
```
FAIL - "AvgPool2dModule_basic"

Unexpected outcome summary:

****** Failed tests - 1 tests
    FAIL - "AvgPool2dModule_basic"
        @ trace item #0 - call to "forward"
        @ output of call to "forward"
        ERROR: value (Tensor with shape=[1, 1, 11, 11] min=+nan, max=+nan, mean=+nan) is not close to golden value (Tensor with shape=[1, 1, 11, 11] min=-0.1026, max=+0.08796, mean=-0.005398)


Summary:
    Failed: 1
```

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 1]
```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
...
+          op, "divisor_override must be either None or a constant int");
+    }
+
+    SmallVector<int64_t, 4> paddingIncludingNC = {0, 0};
+    paddingIncludingNC.insert(paddingIncludingNC.end(), paddingInts.begin(),
+                              paddingInts.end());
+    Value paddedInput = torch_to_linalg::getPaddedTensor(op, rewriter, self,
+                                                         paddingIncludingNC);
+
+    Value N = getDimOp(rewriter, loc, self, 0);
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:298]: this line until line 337 are almost the same as `ConvertAtenMaxPool2dOp`. Can you refactor the code with a helper function?

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 2]
```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
...
+        rewriter.create<linalg::FillOp>(loc, initValue, outTensor).getResult(0);
+
+    auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
+    auto dilationAttr = rewriter.getI64VectorAttr({1, 1});
+    Value windowTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, getAsConstantIndexValues(rewriter, loc, kernelSizeInts),
+        elementType);
+
+    Value sumPool2d = rewriter
+                          .create<linalg::PoolingNhwcSumOp>(
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:338]: The dimension layout required by the linalgOp doesn't match the torch op. We will need to add a new one upstream  https://github.com/llvm/llvm-project/blob/c9c2363048b00749561070008d3b696438ee1024/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L517

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 3]
```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
...
+    SmallVector<int64_t, 2> strideInts;
+    if (!matchPattern(op.stride(), m_TorchConstantIntList(strideInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int strides");
+    SmallVector<int64_t, 2> paddingInts;
+    if (!matchPattern(op.padding(), m_TorchConstantIntList(paddingInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int paddings");
+
+    // The ceil_mode should be a constant `False`.
```
Comment (pashu123) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:263]: Why the `ceil_mode` should be False? Is it a TODO? https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html I see that ceil_mode = True is also supported.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 4]
```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
...
+    bool ceilMode;
+    if (!matchPattern(op.ceil_mode(), m_TorchConstantBool(&ceilMode))) {
+      return rewriter.notifyMatchFailure(op, "ceil_mode must be a constant");
+    } else if (ceilMode) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: ceil_mode is expected to be false");
+    }
+
+    // The count_include_pad should be a constant `True`.
```
Comment (pashu123) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:272]: Same here, the `count_include_pad` can be false. Add this as a TODO.

Comment (pashu123) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:272]: Same here, the `count_include_pad` can be false. Add this as a TODO.

**Review by pashu123 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 2 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: Do you mean to say add a `linalg::PoolingNchwSumOp`? If yes, could you please tell me how to do that or give some references?

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: @cathyzhyi Gentle reminder. I am waiting for your reply on this to go ahead.

Comment (cathyzhyi) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: Sry, I was off yesterday. 
yea, we need to add a linalg.pooling_nchw_sum using Linalg opDSL. You could refer to this [doc page](https://mlir.llvm.org/docs/Dialects/Linalg/OpDSL/) and this PR as a reference https://reviews.llvm.org/D106658.

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: I have added the linalg.pooling_nchw_sum op. https://reviews.llvm.org/D123013

Comment (cathyzhyi) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: looks nice. let's wait for folks that are more familiar with this to review. Does the e2e test pass now with this new op?

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: I haven't tried. I am waiting for it to get merged.

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:338]: > Does the e2e test pass now with this new op?

Yes.

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 1 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:298]: I have multiple PRs opened for the different pooling ops. I will refactor the code in the last remaining PR.

**Comment by vivekkhandelwal1** [comment]
The PR is working fine locally. It has failed here because the `linalg::PoolingNchwSumOp` is not included in the torch-mlir's llvm-project.

**Comment by cathyzhyi** [comment]
FYI, this should now work with #781 merged. Can you rebase to the latest?

**Comment by vivekkhandelwal1** [comment]
> FYI, this should now work with #781 merged. Can you rebase to the latest?

Done.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 1 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Pooling.cpp]: It took a great effort to refactor the code for pooling ops. I hope it looks fine now.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 5]
```diff
@@ -60,30 +56,26 @@ checkAndGetPoolingParameters(OpTy op, ConversionPatternRewriter &rewriter,
...
-                 SmallVectorImpl<int64_t> &kernelSizeInts,
-                 SmallVectorImpl<int64_t> &strideInts,
-                 SmallVectorImpl<int64_t> &paddingInts,
-                 SmallVectorImpl<int64_t> &dilationInts, Value &result) {
+// Creates a pooling operation based on the type specified by `OpTy` and
+// arguments passed.
+template <typename OpTy>
+static LogicalResult createPoolingOp(
+    Operation *op, ConversionPatternRewriter &rewriter, Value self,
+    bool supportFPInput, SmallVectorImpl<int64_t> &kernelSizeInts,
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:64]: nit: `supportFPInput` => `supportFPInputOnly`

**Review by cathyzhyi — APPROVED** [review]
Thanks for taking the effort to refactor the code! Looks good except some nits.

**Pull request merged**